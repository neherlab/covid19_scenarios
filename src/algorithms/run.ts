import * as math from 'mathjs'

import { OneCountryAgeDistribution } from '../assets/data/CountryAgeDistribution.types'

import { SeverityTableRow } from '../components/Main/Scenario/SeverityTable'

import { collectTotals, evolve, getPopulationParams, initializePopulation } from './model'
import { AllParamsFlat, MitigationIntervals } from './types/Param.types'
import { AlgorithmResult, SimulationTimePoint } from './types/Result.types'
import { TimeSeries } from './types/TimeSeries.types'

const identity = (x: number) => x
const poisson = (x: number) => {
  throw new Error('We removed dependency on `random` package. Currently `poisson` is not implemented')
}

interface ChangePoint {
  t: Date
  val: Number[]
}

const compareTimes = (a: ChangePoint, b: ChangePoint): number => {
  return a.t.valueOf() - b.t.valueOf()
}

// NOTE: Assumes containment is sorted ascending in time.
export function interpolateTimeSeries(containment: TimeSeries): (t: Date) => number {
  if (containment.length === 0) {
    return (t: Date) => 1.0
  }
  const Ys = containment.map((d) => d.y)
  const Ts = containment.map((d) => Number(d.t))
  return (t: Date) => {
    if (t <= containment[0].t) {
      return containment[0].y
    }
    if (t >= containment[containment.length - 1].t) {
      return containment[containment.length - 1].y
    }
    const i = containment.findIndex((d) => Number(t) < Number(d.t))

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const evalLinear = (t: number) => {
      const deltaY = Ys[i] - Ys[i - 1]
      const deltaT = Ts[i] - Ts[i - 1]

      const dS = deltaY / deltaT
      const dT = t - Ts[i - 1]

      return Ys[i - 1] + dS * dT
    }

    return evalLinear(Number(t))
  }
}


export function intervalsToTimeSeries(intervals: MitigationIntervals): TimeSeries {
  const changePoints = {}
  intervals.forEach((element) => {
    // bound the value by 0.01 and 100 (transmission can be at most 100 fold reduced or increased)
    const val = Math.min(Math.max(1 - element.mitigationValue, 0.01), 100)

    if (changePoints[element.timeRange.tMin] !== undefined){
      changePoints[element.timeRange.tMin].push(val)
    } else {
      changePoints[element.timeRange.tMin] = [val]
    }
    // add inverse of the value when measure is relaxed
    if (changePoints[element.timeRange.tMax] !== undefined){
      changePoints[element.timeRange.tMax].push(1.0/val)
    } else {
      changePoints[element.timeRange.tMax] = [1.0/val]
    }
  });

  const orderedChangePoints: ChangePoint[] = Object.keys(changePoints).map(d => ({t:new Date(d), val:changePoints[d]}))

  orderedChangePoints.sort(compareTimes)

  if (orderedChangePoints.length) {
    const mitigationSeries: TimeSeries = [{t:orderedChangePoints[0].t, y:1.0}]
    const product = (a: Number,b: Number): Number => (a*b);

    orderedChangePoints.forEach((d,i) => {
      const prevValue = mitigationSeries[2*i].y
      const newValue = d.val.reduce(product, prevValue)
      mitigationSeries.push({t:d.t, y:prevValue})
      mitigationSeries.push({t:d.t, y:newValue})
    })
    return mitigationSeries
  }
  return [];
}

/**
 *
 * Entry point for the algorithm
 *
 */
export async function run(
  params: AllParamsFlat,
  severity: SeverityTableRow[],
  ageDistribution: OneCountryAgeDistribution,
  containment: TimeSeries,
): Promise<AlgorithmResult> {
  const modelParams = getPopulationParams(params, severity, ageDistribution, interpolateTimeSeries(containment))
  const tMin: number = new Date(params.simulationTimeRange.tMin).getTime()
  const tMax: number = new Date(params.simulationTimeRange.tMax).getTime()
  const initialCases = params.suspectedCasesToday
  let initialState = initializePopulation(modelParams.populationServed, initialCases, tMin, ageDistribution)

  function simulate(initialState: SimulationTimePoint, func: (x: number) => number) {
    const dynamics = [initialState]
    let currState = initialState
    while (currState.time < tMax) {
      currState = evolve(currState, modelParams, currState.time + 1, func)
      dynamics.push(currState)
    }

    return collectTotals(dynamics)
  }

  const sim: AlgorithmResult = {
    deterministic: simulate(initialState, identity),
    stochastic: [],
    params: modelParams,
  }

  for (let i = 0; i < modelParams.numberStochasticRuns; i++) {
    initialState = initializePopulation(modelParams.populationServed, initialCases, tMin, ageDistribution)
    sim.stochastic.push(simulate(initialState, poisson))
  }

  return sim
}
