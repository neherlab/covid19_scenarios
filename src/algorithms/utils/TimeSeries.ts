import * as d3 from 'd3'

import { DateRange } from '../types/Param.types'
import { interpolateTimeSeries } from '../run'
import { TimeSeries } from '../types/TimeSeries.types'

export function uniformDatesBetween(min: number, max: number, n: number): Date[] {
  const d = (max - min) / (n - 1)
  const dates = d3.range(min, max + d, d).filter((_, i) => i < n)
  return dates.map((d) => new Date(d))
}

export function makeTimeSeries(simulationTimeRange: DateRange, values: number[]): TimeSeries {
  const { tMin, tMax } = simulationTimeRange
  const n = values.length

  const dates = uniformDatesBetween(tMin.getTime(), tMax.getTime(), n)

  const tSeries = []
  for (let i = 0; i < n; i++) {
    tSeries.push({ t: dates[i], y: values[i] })
  }

  return tSeries
}

export function updateTimeSeries(simulationTimeRange: DateRange, oldTimeSeries: TimeSeries, n: number): TimeSeries {
  const { tMin, tMax } = simulationTimeRange
  const interpolator = interpolateTimeSeries(oldTimeSeries)
  const clamp = function(x: number, min: number, max: number): number {
      if (x < min) {
          return min;
      } else if (x > max) {
          return max;
      } else {
          return x;
      }
  }


  const dates = uniformDatesBetween(tMin.getTime(), tMax.getTime(), n) 
  return dates.map((d) => ({ t: d, y: clamp(interpolator(d), 0, 1.2) }));
}
