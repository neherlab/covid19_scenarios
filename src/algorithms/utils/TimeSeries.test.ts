import { uniformDatesBetween, makeTimeSeries, updateTimeSeries } from './TimeSeries'
import { interpolateTimeSeries } from '../run'

import * as fs from 'fs'
const { utils } = require('jest-snapshot')

expect.extend({
  toBeCloseToArraySnapshot(received: number[]) {
    const { currentTestName, snapshotState } = this
    snapshotState._counters.set(currentTestName, (snapshotState._counters.get(currentTestName) || 0) + 1)
    const count = Number(snapshotState._counters.get(currentTestName))
    const key = utils.testNameToKey(currentTestName, count)
    const expected = snapshotState._snapshotData[key]

    /* If this isn't done, Jest reports the test as 'obsolete' and prompts for deletion. */
    snapshotState._uncheckedKeys.delete(key)

    /* If the snapshot isn't JSON, then return an empty array. Otherwise
     * the test will fails in an inscrutable way.
     */
    const tryParse = (str: string) => {
      try {
        return JSON.parse(str)
      } catch (e) {
        return []
      }
    }

    /* If the expected field is blank, then this is probably a new snapshot. */
    const expectedDeserialized = expected ? tryParse(expected) : []

    const pass =
      expectedDeserialized.length > 0 /* must check this because every() returns true if empty array */
        ? received.every((_, idx) => Math.abs(expectedDeserialized[idx] - received[idx]) < 10 ** -2 / 2)
        : false
    const hasSnapshot = expected !== undefined
    const snapshotIsPersisted = fs.existsSync(snapshotState._snapshotPath)
    const receivedSerialized = JSON.stringify(received, null, 2)

    if (pass) {
      /* Test passed, now update the snapshot state with the serialize snapshot.
       * Technically this is only necessary if no snapshot was saved before. */
      snapshotState._snapshotData[key] = receivedSerialized
    }

    /* This nested mess is derived the Jest snapshot matcher code:
     * https://github.com/facebook/jest/blob/4a59daa8715bde6a1b085ff7f4140f3a337045aa/packages/jest-snapshot/src/State.ts
     */
    if (
      (hasSnapshot && snapshotState._updateSnapshot === 'all') ||
      ((!hasSnapshot || !snapshotIsPersisted) &&
        (snapshotState._updateSnapshot === 'new' || snapshotState._updateSnapshot === 'all'))
    ) {
      if (snapshotState._updateSnapshot === 'all') {
        if (!pass) {
          if (hasSnapshot) {
            snapshotState.updated++
          } else {
            snapshotState.added++
          }
          snapshotState._addSnapshot(key, receivedSerialized, { error: undefined, isInline: false })
        } else {
          snapshotState.matched++
        }
      } else {
        snapshotState._addSnapshot(key, receivedSerialized, { error: undefined, isInline: false })
        snapshotState.added++
      }

      return {
        message: () => 'message a',
        actual: '',
        count,
        expected: '',
        key,
        pass: true,
      }
    } else {
      if (!pass) {
        snapshotState.unmatched++
        return {
          message: () => 'message b',
          actual: receivedSerialized,
          count,
          expected: expected !== undefined ? expected : undefined,
          key,
          pass: false,
        }
      } else {
        snapshotState.matched++
        return {
          message: () => 'message c',
          actual: receivedSerialized,
          count,
          expected: '',
          key,
          pass: true,
        }
      }
    }
  },
})

describe('TimeSeries', () => {
  const tMin = new Date('1970-10-01T00:00:00.000Z')
  const tMax = new Date('1971-02-01T00:00:00.000Z')

  // These are the expected values from uniformDatesBetween(tMin,tMax)
  const tBetween = [
    new Date('1970-10-31T18:00:00.000Z'),
    new Date('1970-12-01T12:00:00.000Z'),
    new Date('1971-01-01T06:00:00.000Z'),
  ]

  describe('uniformDatesBetween()', () => {
    it('returns an array of "n" uniformly spaced dates between "min" and "max", inclusive.', () => {
      const n = 5

      const result = uniformDatesBetween(tMin.getTime(), tMax.getTime(), n)

      expect(result).toBeArray()

      expect(result).toStrictEqual([tMin, ...tBetween, tMax])
    })
  })

  describe('makeTimeSeries()', () => {
    it('maps an array of length "n" values to an array of "n" uniform dates bracketed by the simulation time range.', () => {
      const simulationTimeRange = { tMin, tMax }

      const result = makeTimeSeries(simulationTimeRange, [1, 2, 3, 4, 5])

      expect(result).toBeArray()

      expect(result).toStrictEqual([
        { t: tMin, y: 1 },
        { t: tBetween[0], y: 2 },
        { t: tBetween[1], y: 3 },
        { t: tBetween[2], y: 4 },
        { t: tMax, y: 5 },
      ])
    })
  })

  describe('interpolateTimeSeries()', () => {
    it('if created using an empty vector, the interpolation function must throw', () => {
      expect(() => interpolateTimeSeries([])).toThrow()
    })

    it('if given a time less than tMin, the interpolation function returns the earliest timeseries "y" value.', () => {
      const simulationTimeRange = { tMin, tMax }
      const timeSeries = makeTimeSeries(simulationTimeRange, [1, 3, 2, 5, 4, 7])

      const result = interpolateTimeSeries(timeSeries)

      expect(result).toBeFunction()

      const earlier = new Date(tMin.getTime() - 1000)
      expect(result(earlier)).toBe(1)
    })

    it('if given a time greater than tMax, the interpolation function returns the oldest timeseries "y" value.', () => {
      const simulationTimeRange = { tMin, tMax }
      const timeSeries = makeTimeSeries(simulationTimeRange, [1, 3, 2, 5, 4, 7])

      const result = interpolateTimeSeries(timeSeries)

      expect(result).toBeFunction()

      const later = new Date(tMax.getTime() + 1000)
      expect(result(later)).toBe(7)
    })

    it('interpolation within the timerange occurs as expected.', () => {
      const simulationTimeRange = { tMin, tMax }
      const timeSeries = makeTimeSeries(simulationTimeRange, [1, 7, 10])

      const interpolator = interpolateTimeSeries(timeSeries)

      expect(interpolator).toBeFunction()

      /* Run the interpolator 10 times the [tMin,tMax] range, stepping
       * in intervals of 15 days */

      const intervalCount = 10
      const intervalSize = 15

      const result = [...new Array(intervalCount)].map((_, index) => {
        const date = new Date(Number(tMin))
        date.setDate(date.getDate() + intervalSize * index)
        return interpolator(date)
      })

      expect(result).toBeArrayOfSize(intervalCount)
      expect(result).toBeCloseToArraySnapshot()
    })
  })

  describe('updateTimeSeries()', () => {
    it('interpolates an existing TimeSeries with a new "y" vector.', () => {
      const simulationTimeRange = { tMin, tMax }
      const yVector = [3, 7, 11]
      const timeSeries = makeTimeSeries(simulationTimeRange, yVector)
      const n = 5

      /* Update the time series using the new 'n' value with same TimeRange. */
      const result = updateTimeSeries(simulationTimeRange, timeSeries, n)
      expect(result).toBeArrayOfSize(n).toMatchSnapshot()
    })

    it('interpolates an existing TimeSeries with a new TimeRange.', () => {
      const simulationTimeRange = { tMin, tMax }
      const yVector = [3, 7, 11]
      const timeSeries = makeTimeSeries(simulationTimeRange, yVector)

      /* Add/remove 5 days from tMin and tMax, respectively,
       * resulting in a TimeRange smaller by 10 days.
       */
      const newTMin = new Date(Number(tMin))
      newTMin.setDate(newTMin.getDate() + 5)
      const newTMax = new Date(Number(tMax))
      newTMax.setDate(newTMax.getDate() - 5)
      const newSimulationTimeRange = { tMin: newTMin, tMax: newTMax }

      /* Update the time series using the new TimeRange with same 'n' value */
      const result = updateTimeSeries(newSimulationTimeRange, timeSeries, yVector.length)

      expect(result).toBeArrayOfSize(yVector.length).toMatchSnapshot()
    })
  })
})
