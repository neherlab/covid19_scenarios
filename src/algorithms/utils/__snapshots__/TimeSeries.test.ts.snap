// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeSeries interpolateTimeSeries() interpolation within the timerange occurs as expected. 1`] = `[
  1,
  2.6354594390679185,
  4.209836664611696,
  5.648965995858905,
  6.892037521792783,
  7.889772449043392,
  8.678773876200733,
  9.32431394905387,
  9.891684813683101,
  10
]`;

exports[`TimeSeries updateTimeSeries() interpolates an existing TimeSeries with a new "y" vector. 1`] = `
Array [
  Object {
    "t": 1970-10-01T00:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1970-10-31T18:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1970-12-01T12:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1971-01-01T06:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1971-02-01T00:00:00.000Z,
    "y": 1.2,
  },
]
`;

exports[`TimeSeries updateTimeSeries() interpolates an existing TimeSeries with a new TimeRange. 1`] = `
Array [
  Object {
    "t": 1970-10-06T00:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1970-12-01T12:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1971-01-27T00:00:00.000Z,
    "y": 1.2,
  },
]
`;
